Sources/NIO/EventLoopFuture.swift

```diff
@@ -815,6 +815,37 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that fires only when this `EventLoopFuture` and
+    /// all the provided `futures` complete. It then provides the result of folding the value of this
+    /// `EventLoopFuture` with the values of all the provided `futures`.
+    ///
+    /// This function is suited when you have APIs that already know how to return `EventLoopFuture`s.
+    ///
+    /// The returned `EventLoopFuture` will fail as soon as the a failure is encountered in any of the
+    /// `futures` (or in this one). However, the failure will not occur until all preceding
+    /// `EventLoopFutures` have completed. At the point the failure is encountered, all subsequent
+    /// `EventLoopFuture` objects will no longer be waited for. This function therefore fails fast: once
+    /// a failure is encountered, it will immediately fail the overall EventLoopFuture.
+    ///
+    /// - parameters:
+    ///     - futures: An array of `EventLoopFuture<U>` to wait for.
+    ///     - with: A function that will be used to fold the values of two `EventLoopFuture`s and return a new value wrapped in an `EventLoopFuture`.
+    /// - returns: A new `EventLoopFuture` with the folded value whose callbacks run on `self.eventLoop`.
+    public func fold<U>(_ futures: [EventLoopFuture<U>], with combiningFunction: @escaping (T, U) -> EventLoopFuture<T>) -> EventLoopFuture<T> {
+        let body = futures.reduce(self) { (f1: EventLoopFuture<T>, f2: EventLoopFuture<U>) -> EventLoopFuture<T> in
+            let newFuture = f1.and(f2).then({ (args: (T, U)) -> EventLoopFuture<T> in
+                let (f1Value, f2Value) = args
+                assert(self.eventLoop.inEventLoop)
+                return combiningFunction(f1Value, f2Value)
```

Comment:
Do we really need a `hopTo` before we return the `EventLoopFuture` from the `combiningFunction`? I think `then` accepts a `callback` whose returned `future` runs on any `eventLoop` and executes this `callback` on its caller's `eventLoop` and always returns a future on its caller's `eventLoop` (the result of `and`) which already (because of `and`) `hopTo` its caller's `eventLoop`, so if we start the reduce with `self`, it will always return a `future` whose callbacks run on `self.eventLoop`.

Reply:
Ah, you're right, the `EventLoopFuture` returned from `then` does not need to be hopped: it is tied to the loop of `f1`. The assert is sufficient.

Want to remove the parentheses from the `then` call? They shouldn't be needed.

Reply:
Ah yes